### PR TITLE
TA emulation latencies

### PR DESCRIPTION
**Pull request by ArturSztuc**
Expanding TA latencies measurements for the emulation (`process_tpstream`) with an extra script that converts the output latencies `.csv` format to plotting-ready one. New documentation added.

Changes:
1. `process_tpstream` now has an option `--latencies`, which, if enabled, will cause generation of a large `csv` file with times taken to process each TP into TA by the emulator (and another file with times to process each TA onto TC). Each column represents a TP, and rows represent; time_start of the TP; adc_integral of the TP; time taken to process the TP by TAMaker; whether the TP resulted in TA made (1) or not (0);
2. New `convert_tplatencies.py` script which takes the above `.csv` together with the output `hdf5` file and creates another `.csv` with latencies per-TA level Each row represents a TA made, and columns represent; time to process the TP that caused TA creation; time to process TPs from the first TP in TA window to the TP that caused TA creation; time to process all the TPs in a TA window; mean time to process a TP, extracted from the TPs in a TA window; number of TPs in a TA.
3. Update to README.md, & new readme for the ta-processing script.

The output of  `convert_tplatencies.py` can be very easily used to make latencies plots for TAs as function of number of TPs in a TA. E.g. with script like below:

```python
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib import rc

csv_files = [ "latencies_simplified_channeladjacency_medium.csv", "latencies_simplified_channeladjacency_long.csv",
              "latencies_simplified_channeldistance_medium.csv", "latencies_simplified_channeldistance_long.csv",
              "latencies_simplified_horizontalmuon_medium.csv", "latencies_simplified_horizontalmuon_long.csv",
              "latencies_simplified_DBSCAN_medium.csv"]

colors = ['darkgreen', 'palegreen', 'darkblue', 'skyblue', 'darkred', 'lightcoral', 'darkorange']

labels = ['ChAdj S', 'ChAdj L', 'ChDist S', 'ChDist L', 'HMA S', 'HMA L', 'DBSCAN M']

axislabels = ['Processing time for the last TP producing a TA',
              'Processing time TPs from window start to TP producing TA',
              'Processing time for all the TPs in a TA window',
              'Mean time taken to process the TPs in a TA window']

outputnames = ['latencies_lasttp.png', 'latencies_sumtolasttp.png', 'latencies_sumtawindow.png', 'latencies_meantawindow.png']

for measurementtype in range(len(axislabels)):
    fig, ax = plt.subplots()
    print(f'Generating plot titled: "{axislabels[measurementtype]}"')

    for i, file in enumerate(csv_files):
        df = pd.read_csv(file)
        ax.scatter(df.iloc[:, 4], df.iloc[:, measurementtype], label=labels[i], lw=0, marker=',', color=colors[i], s=1, alpha=0.4)

    ax.set_ylabel('Time (ns)', fontsize=14)
    ax.set_xlabel('Number of TPs in TA', fontsize=14)
    ax.set_yscale('log')
    ax.set_xscale('log')
    ax.grid(True)
    ax.tick_params(axis='both', which='major', labelsize=10)
    ax.legend(fontsize=10, ncols=4, columnspacing=0.1)

    plt.title(axislabels[measurementtype],fontsize=15)

    plt.tight_layout()
    plt.savefig(outputnames[measurementtype], dpi=300)
 ```

I got these plots:
latencies_lasttp
latencies_sumtolasttp
latencies_sumtawindow
latencies_meantawindow

### CODE REVIEW

**Comment by alessandrothea** [comment]
Hi, the plots look cool!

A couple of basic questions, just to warm up:

1. Does the `.csv` file has really a column per TP? not a row? Seems a bizarre choice
2. What is the definition of "times taken to process each TP into TA". It doesn't seem obvious to me.

**Comment by ArturSztuc** [comment]
@alessandrothea 
> 1. Does the `.csv` file has really a column per TP? not a row? Seems a bizarre choice

It doesn't need to be this way. But for each TP we have 4 values (start_time, adc_integral, latency, boolean if this TP resulted in TA being made), so it would be 4 rows per TP, and it would make it difficult to just look at that csv file raw.

> 2. What is the definition of "times taken to process each TP into TA". It doesn't seem obvious to me.

Do you mean each TP in TA window? So first we get the time taken to process a TP by a TAMaker here: https://github.com/DUNE-DAQ/trgtools/blob/18779b15fbf0a88c91df6170aae17795fd1ca213/include/trgtools/detail/EmulationUnit.hxx#L103-L107
and that's what's being saved in the large `.csv` output, together with some other info per-TP to e.g. be able to identify it later when comparing against output `.hdf5`

In the post-processing python script, one of the measurements is the "time taken to process ALL the TPs in a TA window": each TA has some window defined by the first TP's `start_time` and last TP's `start_time` (although that could be somewhat algo-dependent), and we simply find the first TP from CSV file in that window, last TP in that window, and sum the latencies together.

**Comment by ArturSztuc** [comment]
@MRiganSUSX and @aeoranday could you please review this PR? 
It's been almost a week, and I don't want to work over the night making last-minute changes before the deadline again...

File: scripts/convert_tplatencies.py
Comment (aeoranday) [inline scripts/convert_tplatencies.py:1]: This file should have `chmod +x`.

File: scripts/convert_tplatencies.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    tp_data_latencies = pd.read_csv(input_latencies, header=None)
+    
+    # Pre-allocate the output latencies
+    output_array = np.zeros((len(ta_data.tp_data), 5), dtype=np.uint64)
+
+    # Find all the TPs that caused creation of a TA
+    tpidx_making_tas = np.where(tp_data_latencies.iloc[:, 3].values == 1)
+
+    last_tpidx_making_tas = 0
+    changed_tpidx_making_tas = 0
```
Comment (aeoranday) [inline scripts/convert_tplatencies.py]: `changed_tp_idx_making_tas` is not used.

**Review by aeoranday — CHANGES_REQUESTED** [review]
Just a few small changes to `convert_tplatencies.py`. Everything from the documentation and C++ implementation looks good otherwise!

File: scripts/convert_tplatencies.py
Comment (ArturSztuc) [reply scripts/convert_tplatencies.py:1]: Thanks, chmoded!

File: scripts/convert_tplatencies.py (hunk 1 above)
Comment (ArturSztuc) [reply scripts/convert_tplatencies.py]: Good spot, cheers. I also found another example of that, removed both.

**Review by aeoranday — APPROVED** [review]
LGTM. It would be useful to have a dedicated plotting example for these latencies (as in this PR's description) sometime in the future.

**Pull request merged**